services/src/unit-test/java/uk/gov/moj/sdt/services/TargetApplicationSubmissionServiceTest.java

```diff
@@ -307,225 +394,212 @@ public Object answer() throws Throwable {
-        connectionTimeOutParam.setName("TARGET_APP_TIMEOUT");
+        connectionTimeOutParam.setName(TARGET_APP_TIMEOUT);
         connectionTimeOutParam.setValue("1000");
-        EasyMock.expect(this.mockCacheable.getValue(IGlobalParameter.class, "TARGET_APP_TIMEOUT")).andReturn(
+        when(this.mockCacheable.getValue(IGlobalParameter.class, TARGET_APP_TIMEOUT)).thenReturn(
                 connectionTimeOutParam);
 
         final IGlobalParameter receiveTimeOutParam = new GlobalParameter();
-        receiveTimeOutParam.setName("TARGET_APP_RESP_TIMEOUT");
-        receiveTimeOutParam.setValue("12000");
-        EasyMock.expect(this.mockCacheable.getValue(IGlobalParameter.class, "TARGET_APP_RESP_TIMEOUT")).andReturn(
+        receiveTimeOutParam.setName(TARGET_APP_RESP_TIMEOUT);
+        receiveTimeOutParam.setValue(TWELVE_THOUSAND);
+        when(this.mockCacheable.getValue(IGlobalParameter.class, TARGET_APP_RESP_TIMEOUT)).thenReturn(
                 receiveTimeOutParam);
 
         final TimeoutException timeoutEx = new TimeoutException("Timeout occurred", "Timeout occurred");
         this.mockConsumerGateway.individualRequest(individualRequest, 1000, 12000);
-        EasyMock.expectLastCall().andThrow(timeoutEx);
+        doThrow(timeoutEx).when(mockConsumerGateway).individualRequest(individualRequest, 1000, 12000);
 
         final IErrorMessage errorMsg = new ErrorMessage();
         errorMsg.setErrorCode("REQ_NOT_ACK");
         errorMsg.setErrorDescription("Request not acknowledged");
         errorMsg.setErrorText("Request Not Acknowledged");
 
-        EasyMock.expect(this.mockErrorMsgCacheable.getValue(IErrorMessage.class, "REQ_NOT_ACK")).andReturn(errorMsg);
+        when(this.mockErrorMsgCacheable.getValue(IErrorMessage.class, "REQ_NOT_ACK")).thenReturn(errorMsg);
 
         // Now create an ErrorLog object with the ErrorMessage object and the IndividualRequest object
         final IErrorLog errorLog = new ErrorLog(errorMsg.getErrorCode(), errorMsg.getErrorText());
 
         individualRequest.setErrorLog(errorLog);
 
-        this.mockIndividualRequestDao.persist(individualRequest);
-        EasyMock.expectLastCall();
-
         final IGlobalParameter maxForwardingAttemptsParam = new GlobalParameter();
         maxForwardingAttemptsParam.setName("MAX_FORWARDING_ATTEMPTS");
         maxForwardingAttemptsParam.setValue("3");
-        EasyMock.expect(this.mockCacheable.getValue(IGlobalParameter.class, "MAX_FORWARDING_ATTEMPTS")).andReturn(
+        when(this.mockCacheable.getValue(IGlobalParameter.class, "MAX_FORWARDING_ATTEMPTS")).thenReturn(
                 maxForwardingAttemptsParam);
 
-        this.mockMessageWriter.queueMessage(EasyMock.isA(ISdtMessage.class), EasyMock.isA(String.class),
-                EasyMock.anyBoolean());
-        EasyMock.expectLastCall();
+        this.mockMessageWriter.queueMessage(isA(ISdtMessage.class), isA(String.class), anyBoolean());
```

Comment:
Think this needs to be wrapped in a doAnswer().

Reply:
Need this to be explained.

Reply:
This line is a hang over from the way EasyMock handled mocking.  Under EasyMock each mocked object starts off in _record_ mode.  You set up exactly what you expect to happen to each mocked object during the test and then set them to _replay_ mode.  Finally, you _verify_ each mocked object to confirm that what was expected to happen happened.

This line is an example of one that was used in EasyMock to set up a mocked object in _record_ mode to expect a void method to be called on it (in conjunction with ```expectLastCall()```).  A bit later on the object would be switched to _replay_ mode so that the test could be carried out.

Mockito doesn't have the concept of separate _record_ and _replay_ modes - the mocked objects remember all methods that are directly called on them all the time.  To distinguish between a method being called on a mocked object and a method expecting to be called on one you usually wrap the method call in ```Mockito.when()```, e.g. ```Mockito.when(mockMyObject.getName()).thenReturn("Bob");```.  You only use this for methods that return something though.

For methods that don't return a value you can wrap them in alternative Mockito methods, depending on what you expect to happen.  For example, if you want to throw an exception you'd wrap it ```Mockito.doThrow()```.  If you just expect the void method to be called normally and don't need to check anything else about the call you can just use ```Mockito.verify()``` at the end of the test to confirm that is has been called.

Under EasyMock this line set up the mock ```MessageWriter``` object to expect the ```queueMessage()``` method to be called on it, but under Mockito it's actually calling the ```queueMessage()``` method on it.  This could lead to unexpected results as Mockito reports that a particular method has been called more times than expected.

Here it looks as though all we really want to check is that ```queueMessage()``` was called.  Rather than wrapping it in something like ```doAnswer()``` as originally thought, all we need to do is change the line to ```Mockito.verify(mockMessageWriter).queueMessage(...)``` and move it after ```targetAppSubmissionService.processRequestToSubmit()``` has been called.